Handle Russian and Ukrainian plurals

bazzadp opened the pull request:
Fixes #1968 using [plural-ru](https://github.com/andrewp-as-is/plural-ru.py) function. Note this is not available as an importable function so added code directly to our code base (checked the licence first).

Home page intro:
![Home page intro](https://user-images.githubusercontent.com/10931297/106815641-11703700-666c-11eb-9b2f-a0224ab2a3d2.png)

Methodology text:
![Methodology text](https://user-images.githubusercontent.com/10931297/106815698-2056e980-666c-11eb-83f2-930bd5663761.png)

Contributors:
![Contributors](https://user-images.githubusercontent.com/10931297/106816491-5e084200-666d-11eb-8401-cd3c2ebe3c16.png)

@dtikhonov can you confirm I've used the right words?

English word|ends in 1 (except 11)| ends in 2,3 and 4 (except 12, 13, 14)|else
-----|-----|----|----
chapters|глава|главы|глав
contributors|участник|участника|участников
people|человек|человека|человек

It does look like we used "главах" below but Google Translate didn't come up with that at all so I guess double check that?

@tymosh does Ukrainian follow the same rules? Should we do the same for those templates?

tymosh commented:
Hi @bazzadp. Yes, the same mechanism works for Ukrainian:

English word|ends in 1 (except 11)| ends in 2,3 and 4 (except 12, 13, 14)|else
-----|-----|----|----
chapters|розділ|розділи|розділів
contributors|контриб'ютор|контриб'ютори|контриб'юторів
people|людина|людини|людей

I will also need to check sentences for correct words endings in context.

bazzadp commented:
Thanks @tymosh . Added those in [the last commit](https://github.com/HTTPArchive/almanac.httparchive.org/pull/1969/commits/f091286bdccdff85d123ca7e5ecd6ae1759b0c30) if you wanna double check.

Also added test site here: https://20210204t112250-dot-webalmanac.uk.r.appspot.com

tymosh commented on src/templates/uk/base.html:
```diff
@@ -20,7 +20,7 @@
 {% block mission %}
 <p>
-Наша місія - поєднати суху статистику і тренди від HTTP Archive з досвідом веб-спільноти. Web Almanac це всеосяжний звіт про стан вебу, підтриманий реальними даними та довіреними веб-експертами. Видання {{ year }} року складається з {{ supported_chapters | length }} розділів, які охоплюють аспекти контенту сторінки, досвіду користування, публікації та розповсюдження.
+Наша місія - поєднати суху статистику і тренди від HTTP Archive з досвідом веб-спільноти. Web Almanac це всеосяжний звіт про стан вебу, підтриманий реальними даними та довіреними веб-експертами. Видання {{ year }} року складається з {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділ", "розділи", "розділів"]) }}, які охоплюють аспекти контенту сторінки, досв ...
```
```suggestion
Наша місія - поєднати суху статистику і тренди від HTTP Archive з досвідом веб-спільноти. Web Almanac - це всеосяжний звіт про стан вебу, підтриманий реальними даними та довіреними веб-експертами. Видання {{ year }} року складається з {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділу", "розділів", "розділів"]) }}, які охоплюють аспекти контенту сторінки, досвіду користування, публікації та розповсюдження.
```

tymosh commented on src/templates/uk/base.html:
```diff
@@ -117,7 +117,7 @@
 {% block methodology_link %}Дізнатися про нашу Методологію{% endblock %}
 {% block methodology_description %}
-Якщо не вказано інше, метрики в усіх {{ supported_chapters | length }} розділах Web Almanac {{ year }} взяті з датасету HTTP Archive. HTTP Archive – це громадський проект, що відслідковує, як будується веб з 2010 року. Метадата про {{ self.total_websites() }} вебсайтів тестується щомісяця, використовуючи WebPageTest та Lighthouse, і додається до публічної бази даних BigQuery для аналізу. Для метр ...
+Якщо не вказано інше, метрики в усіх {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділ", "розділи", "розділів"]) }} Web Almanac {{ year }} взяті з датасету HTTP Archive. HTTP Archive – це громадський проект, що відслідковує, як будується веб з 2010 року. Метадата про {{ self.total_websites() }} вебсайтів тестується щомісяця, використовуючи WebPageTest та L ...
```
```suggestion
Якщо не вказано інше, метрики в усіх {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділі", "розділах", "розділах"]) }} Web Almanac {{ year }} взяті з датасету HTTP Archive. HTTP Archive – це громадський проект, що відслідковує, як будується веб з 2010 року. Метадата про {{ self.total_websites() }} вебсайтів тестується щомісяця, використовуючи WebPageTest та Lighthouse, і додається до публічної бази даних BigQuery для аналізу. Для метрик Web Almanac {{ year }} був використаний як базис набір даних від {{ self.dataset() }} року. Більше інформації на сторінці Методологія.
```

tymosh commented on src/templates/uk/2019/contributors.html:
```diff
@@ -2,9 +2,9 @@
 {% block title %}Контриб'ютори 2019 року | Web Almanac від HTTP Archive{% endblock %}
 
-{% block description %}{{ config.contributors.items() | length }} неймовірних людей, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
+{% block description %}{{ config.contributors.items() | length }} неймовірних {{ plural_ru(config.contributors.items() | length | int, ["людина", "людини", "людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```
```suggestion
{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```

tymosh commented on src/templates/uk/2020/contributors.html:
```diff
@@ -2,9 +2,9 @@
 {% block title %}Контриб'ютори {{ year }} року | Web Almanac від HTTP Archive{% endblock %}
 
-{% block description %}{{ config.contributors.items() | length }} неймовірних людей, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
+{% block description %}{{ config.contributors.items() | length }} неймовірних {{ plural_ru(config.contributors.items() | length | int, ["людина", "людини", "людей"]) }},, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```
```suggestion
{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```

tymosh reviewed (changes requested):
Here's another thing: there are different word endings for "contributors" in constructions "124 contributors" and "5 of 124 contributors" in Ukrainian. Is there any way to do that?

dtikhonov commented on src/templates/ru/2020/contributors.html:
```diff
@@ -2,9 +2,9 @@
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["человек", "человека", "человек"]) }}, внесших свой вклад в Web Almanac за {{ year }} год — это аналитики, авторы, дизайнеры, разработчики, редакторы, руководители, проверяющие и переводчики.{% endblock %}
 
-{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> участников.{% endblock %}
+{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["участник", "участника", "участников"]) }}.{% endblock %}
```
You need genitive here:
> "участника", "участников", "участников"

dtikhonov commented on src/templates/ru/base.html:
```diff
@@ -20,7 +20,7 @@
 {% block mission %}
 <p>
-Наша задача — объединить необработанные статистические данные и тенденции HTTP Archive вместе со знаниями веб-сообщества. Web Almanac — это всеобъемлющий отчёт о состоянии веба, подкреплённый реальными данными и подтверждённый авторитетными веб-экспертами. Издание {{ year }} года состоит из {{ supported_chapters | length }} глав, охватывающих такие аспекты, как содержание страниц, пользовательски ...
+Наша задача — объединить необработанные статистические данные и тенденции HTTP Archive вместе со знаниями веб-сообщества. Web Almanac — это всеобъемлющий отчёт о состоянии веба, подкреплённый реальными данными и подтверждённый авторитетными веб-экспертами. Издание {{ year }} года состоит из {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["глава", "главы", "глав" ...
```
Same here:
> "главы", "глав", "глав"

dtikhonov commented on src/templates/ru/base.html:
```diff
@@ -117,7 +117,7 @@
 {% block methodology_link %}Узнать подробнее про методологию{% endblock %}
 {% block methodology_description %}
-Если не указано иное, метрики во всех {{ supported_chapters | length }} главах Web Almanac {{ year }} получены из набора данных HTTP Archive. HTTP Archive — это управляемый сообществом проект по исследованию веба с 2010 года. При помощи WebPageTest и Lighthouse ежемесячно проверяются метаданные {{ self.total_websites() }} сайтов, которые затем вносятся в общедоступную базу данных BigQuery для пос ...
+Если не указано иное, метрики во всех {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["глава", "главы", "глав"]) }} Web Almanac {{ year }} получены из набора данных HTTP Archive. HTTP Archive — это управляемый сообществом проект по исследованию веба с 2010 года. При помощи WebPageTest и Lighthouse ежемесячно проверяются метаданные {{ self.total_websites() }} сай ...
```
Here, the translation does not consider that there may be just one chapter: the word всех (all) cannot be used with 1!  (E.g. "in all one chapters"... does not make sense.)

Nevertheless, I think we can safely ignore the possibility that the Almanac has only one chapter.

The forms of the words you want are
> "глава", "главах", "главах"

dtikhonov commented on src/templates/ru/2019/contributors.html:
```diff
@@ -2,9 +2,9 @@
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["человек", "человека", "человек"]) }}, внесших свой вклад в Web Almanac за 2019 год — это аналитики, авторы, участники мозговых штурмов, дизайнеры, разработчики, редакторы, руководители, проверяющие и переводчики.{% endblock %}
 
-{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> участников.{% endblock %}
+{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["участник", "участника", "участников"]) }}.{% endblock %}
```
> "участника", "участников", "участников"

dtikhonov commented on src/templates/uk/2019/contributors.html:
```diff
@@ -2,9 +2,9 @@
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}.{% endblock %}
```
You missed a couple, @tymosh

bazzadp replied:
Resolving this as code has changed so let's pick up on latest code by doing another review (and ideally suggesting the changes using the ![suggest](https://user-images.githubusercontent.com/10931297/106916108-b89dae80-66fe-11eb-8204-2b3592a29180.png) icon for my poor, non-Russian/non-Ukranian brain!).

dtikhonov commented on src/templates/uk/2020/contributors.html:
```diff
@@ -2,9 +2,9 @@
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}.{% endblock %}
```
and this one.

bazzadp replied:
Resolving this as code has changed so let's pick up on latest code by doing another review (and ideally suggesting the changes using the ![suggest](https://user-images.githubusercontent.com/10931297/106916108-b89dae80-66fe-11eb-8204-2b3592a29180.png) icon for my poor, non-Russian/non-Ukranian brain!).

bazzadp commented:
> Thanks! Here's another thing: there are different word endings for "contributors" in constructions "124 contributors" and "5 of 124 contributors" in Ukrainian. Is there any way to do that?

Yes and have this working now. Take a look at the latest code. Can also view it here: https://20210204t153224-dot-webalmanac.uk.r.appspot.com

@dtikhonov I presume this is the same in Russian? Can you review the latest code?

tymosh commented on src/templates/uk/2020/contributors.html:
```diff
@@ -2,9 +2,9 @@
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span><span id="contributors-t ...
```
```suggestion
{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютора", "контриб'юторів", "контриб'юторів"]) }}</span><span id="contributors-total-text-non-filtered"> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span>.{% endblock %}
```

tymosh commented on src/templates/uk/2019/contributors.html:
```diff
@@ -2,9 +2,9 @@
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span><span id="contributors-t ...
```
```suggestion
{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютора", "контриб'юторів", "контриб'юторів"]) }}</span><span id="contributors-total-text-non-filtered">{{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span>.{% endblock %}
```

tymosh reviewed (approved):
Looks good!
